### PR TITLE
docker role and playbook

### PR DESCRIPTION
**Pull request by NakulManchanda**
Signed-off-by: Nakul Manchanda <nakul.manchanda@ibm.com>

- PR has following changes:

    **Evironment Vars**
    - it introduce some new variables
       ZOWE_EXTERNAL_HOST - how to access zowe system externally
       ZOWE_ZOS_HOST - where is zos machine installed
       ZOWE_ZOSMF_HOST
       ZOWE_ZSS_HOST
      existing ZOWE_EXPLORER_HOST - this is different from all obove
    - It adds new test scope "docker test" to test docker image         
       
    **Playbook & Roles**   
    - new "docker" role
    - install-docker.yml - new docker install
    - add docker logs to show-logs.yml playbook

    **Sanity test** 
    - requires new environment variable ZOWE_EXTERNAL_HOST
    - eg docker - ZOWE_EXTERNAL_HOST=localhost, ZOWE_ZOS_HOST=marist-1, or any other target server
    - `npm run test:docker` - new npm script to run sanity test against local docker image
    - Sanity test refactor to replace SSH_HOST with ZOWE_EXTERNAL_HOST

      
    **Readme's**
     - updated docker readme
     - updated sanity test readme

Please check if your PR fulfills the following requirements. This is simply a reminder of what we are going to look for before merging your PR. If you don't know all of this information when you create this PR, don't worry. You can edit this template as you're working on it.

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Necessary documentation (if appropriate) have been added / updated
- [ ] DCO signoffs have been added to all commits, including this PR

#### PR type
What type of changes does your PR introduce to Zowe? _Put an `x` in the box that applies to this PR. If you're unsure about any of them, don't hesitate to ask._ 
- [ ] Bugfix 
- [ ] Feature 
- [ ] Other... Please describe:

#### Relevant issues

Fixes 

#### Changes proposed in this PR

- 

#### Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

#### Does this PR do something the person installing Zowe should know about?

---
* Affected function: _general area of interest_ *

---
* Description: _1 line description_ *

---
* Part: _name of customizable file involved_  *

---
_multi-line description_

#### Is there a related doc issue or Pull Request? 

Doc issue/PR number:

### CODE REVIEW

File: playbooks/install-docker.yml
Comment (stevenhorsman) [inline playbooks/install-docker.yml:78]: should this be zss, or do you need to launch the app server on z/OS as well?

File: tests/sanity/README.md
Comment (stevenhorsman) [inline tests/sanity/README.md:104]: Is ZOWE_ZOS_HOST used anywhere?

File: playbooks/install-docker.yml
Comment (NakulManchanda) [reply playbooks/install-docker.yml]: modified

File: tests/sanity/README.md
Comment (NakulManchanda) [reply tests/sanity/README.md]: removed - not used

File: playbooks/install-docker.yml
Comment (NakulManchanda) [inline playbooks/install-docker.yml:44]: @jackjia-ibm one suggestion lets move this configure to docker role

Comment (NakulManchanda) [reply playbooks/install-docker.yml:44]: docker>
  tasks>
     configure-zowe-for-docker.yml

Comment (NakulManchanda) [reply playbooks/install-docker.yml]: add configure and uninstall to docker role

File: playbooks/roles/configure/defaults/main.yml
Comment (jackjia-ibm) [inline playbooks/roles/configure/defaults/main.yml:112]: We had split api-mediation into 4 packages. Please update this default value to empty string. I believe this way, we can make it compatible with staging, and also won't conflict the value we may change in the future. For example, adding caching-service to the list.

File: playbooks/roles/docker/defaults/main.yml
Comment (jackjia-ibm) [inline playbooks/roles/docker/defaults/main.yml:23]: I believe this value doesn't need to be here since the verify playbook/role won't read this config file.

File: playbooks/roles/docker/tasks/docker-logs.yml
Comment (jackjia-ibm) [inline playbooks/roles/docker/tasks/docker-logs.yml:15]: This will show log of the first docker container. Maybe we can filter from the image name which should give you better chance to find the correct container.

BTW, with test run, do you find `container_id` could be empty?

Comment (jackjia-ibm) [reply playbooks/roles/docker/tasks/docker-logs.yml:15]: Does `docker ps -a -q --filter ancestor=` give you failure? If so, you may also need to fix `playbooks/roles/docker/tasks/uninstall.yml` since it's using similar command.

File: playbooks/roles/docker/tasks/uninstall.yml
Comment (jackjia-ibm) [inline playbooks/roles/docker/tasks/uninstall.yml:4]: Does this work well?

File: playbooks/roles/start/tasks/wait_for_zowe.yml
Comment (jackjia-ibm) [inline playbooks/roles/start/tasks/wait_for_zowe.yml:92]: Please change to `when: zowe_zos_host == zowe_external_domain_name and ("gateway" in zowe_launch_components or zowe_launch_components == "")`

File: playbooks/roles/start/tasks/wait_for_zowe.yml
Comment (jackjia-ibm) [inline playbooks/roles/start/tasks/wait_for_zowe.yml:114]: Please change to `when: zowe_zos_host == zowe_external_domain_name and ("api-catalog" in zowe_launch_components or zowe_launch_components == "")`

File: playbooks/roles/start/tasks/wait_for_zowe.yml
Comment (jackjia-ibm) [inline playbooks/roles/start/tasks/wait_for_zowe.yml:138]: Please change to `when: zowe_zos_host == zowe_external_domain_name and ("app-server" in zowe_launch_components or zowe_launch_components == "")`

File: playbooks/roles/start/tasks/wait_for_zowe.yml
Comment (jackjia-ibm) [inline playbooks/roles/start/tasks/wait_for_zowe.yml:75]: Please change to `when: zowe_zos_host == zowe_external_domain_name and ("jobs-api" in zowe_launch_components or zowe_launch_components == "")`

File: playbooks/roles/start/tasks/wait_for_zowe.yml
Comment (jackjia-ibm) [inline playbooks/roles/start/tasks/wait_for_zowe.yml:58]: Please change to `when: zowe_zos_host == zowe_external_domain_name and ("app-server" in zowe_launch_components or zowe_launch_components == "")`

File: tests/installation/src/constants.ts
Comment (jackjia-ibm) [inline tests/installation/src/constants.ts:35]: I think this value is not used any more? Or do you plan to use it in `tests/installation/src/__tests__/basic/install-docker.ts`?

**Review by jackjia-ibm — CHANGES_REQUESTED** [review]
Thanks Nakul. There are few changes mostly related to the `zowe_launch_components` value

**Comment by jackjia-ibm** [comment]
Please note, after you merge https://github.com/zowe/zowe-dockerfiles/pull/21, `Dockerfile.jenkins` will be removed. You will also need to change `Jenkinsfile` `docker build -f Dockerfile.jenkins` commands to use correct Dockerfile.

File: playbooks/roles/docker/tasks/docker-logs.yml
Comment (NakulManchanda) [reply playbooks/roles/docker/tasks/docker-logs.yml:15]: container id exists , we call that from  `install-docker.yml`

empty case - when we evoke it from `show-logs.yml`  as sometimes we quite before reaching install-docker log
,at that level we have zowe_docker_url only not even zowe_docker_image and zowe_image_tag
, so kept it without filters

File: playbooks/roles/docker/tasks/uninstall.yml
Comment (NakulManchanda) [reply playbooks/roles/docker/tasks/uninstall.yml:4]: yes, tried many times locally works great, as far as we are removing same image and tag

File: tests/installation/src/constants.ts
Comment (NakulManchanda) [reply tests/installation/src/constants.ts:35]: no plans, i can remove it

File: playbooks/roles/docker/tasks/uninstall.yml
Comment (jackjia-ibm) [reply playbooks/roles/docker/tasks/uninstall.yml:4]: what happens in the case you mentioned in another block: at that level we have zowe_docker_url only not even zowe_docker_image and zowe_image_tag?

File: playbooks/roles/docker/defaults/main.yml
Comment (NakulManchanda) [reply playbooks/roles/docker/defaults/main.yml]: removed

File: playbooks/roles/configure/defaults/main.yml
Comment (NakulManchanda) [reply playbooks/roles/configure/defaults/main.yml]: cleared value

File: playbooks/roles/docker/tasks/uninstall.yml
Comment (NakulManchanda) [reply playbooks/roles/docker/tasks/uninstall.yml:4]: it wont break pipeline `ignore_errors: True`
and help feature to re-run same playbook with same param locally

File: playbooks/roles/docker/tasks/docker-logs.yml
Comment (NakulManchanda) [reply playbooks/roles/docker/tasks/docker-logs.yml:15]: In pipeline, 

Called here:
https://github.com/zowe/zowe-install-packaging/blob/b01af5d55ed4030ebd6192fbb899e94d53b07b4f/tests/installation/src/__tests__/basic/install-docker.ts#L45

Evoked Here:
https://github.com/zowe/zowe-install-packaging/blob/b01af5d55ed4030ebd6192fbb899e94d53b07b4f/playbooks/show-logs.yml#L26

File: playbooks/roles/docker/tasks/uninstall.yml
Comment (NakulManchanda) [reply playbooks/roles/docker/tasks/uninstall.yml:4]: we can always add additional ways to clear running container based on different conditionals and situation we encounter and enhance it

**Comment by NakulManchanda** [comment]
@stevenhorsman @jackjia-ibm 
- changes look good now, lets merge this tomm,
- got thorough review from @jackjia-ibm - addressed all feedback based on alternatives provided

- it passed smpe successfully
- 160 pass 1 fail in convenience
- docker expected failure as image don't have zss and caching service changes

Overall, seems ready to merge, after last review and changes from jack

File: playbooks/roles/docker/tasks/uninstall.yml
Comment (NakulManchanda) [reply playbooks/roles/docker/tasks/uninstall.yml]: modified as per query provided @jackjia-ibm

File: playbooks/roles/docker/tasks/docker-logs.yml
Comment (NakulManchanda) [reply playbooks/roles/docker/tasks/docker-logs.yml]: modified as per query provided @jackjia-ibm

**Review by jackjia-ibm — APPROVED** [review]
Looks good to me. Thanks a lot Nakul!

**Review by stevenhorsman — APPROVED** [review]

**Pull request merged**
